### PR TITLE
JAMES-2756 SimpleMessageSearchIndex doesn't decode header values

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

**Comment by trantienduchn** [comment]
There are some other fails on the SimpleMessageSearchIndex tests. And I think this modification explore another bug :)

**Comment by chibenwa** [comment]
Rebased to pass unit tests

**Comment by chibenwa** [comment]
```
[d578995d125de5e30aadcd96a9e6010d36251651] [ERROR] Failures: 
[d578995d125de5e30aadcd96a9e6010d36251651] [ERROR]   SimpleMessageSearchIndexTest>AbstractMessageSearchIndexTest.addressShouldReturnUidHavingRightExpeditorWhenFromIsSpecifiedWithDomainPartOfEmail:819 
[d578995d125de5e30aadcd96a9e6010d36251651] Expecting:
[d578995d125de5e30aadcd96a9e6010d36251651]   <[]>
[d578995d125de5e30aadcd96a9e6010d36251651] to contain only:
[d578995d125de5e30aadcd96a9e6010d36251651]   <[MessageUid{uid=8}]>
[d578995d125de5e30aadcd96a9e6010d36251651] but could not find the following elements:
[d578995d125de5e30aadcd96a9e6010d36251651]   <[MessageUid{uid=8}]>
[d578995d125de5e30aadcd96a9e6010d36251651] 
[d578995d125de5e30aadcd96a9e6010d36251651] [ERROR]   SimpleMessageSearchIndexTest>AbstractMessageSearchIndexTest.addressShouldReturnUidHavingRightExpeditorWhenFromIsSpecifiedWithOnlyUserPartOfEmail:807 
[d578995d125de5e30aadcd96a9e6010d36251651] Expecting:
[d578995d125de5e30aadcd96a9e6010d36251651]   <[]>
[d578995d125de5e30aadcd96a9e6010d36251651] to contain only:
[d578995d125de5e30aadcd96a9e6010d36251651]   <[MessageUid{uid=8}]>
[d578995d125de5e30aadcd96a9e6010d36251651] but could not find the following elements:
[d578995d125de5e30aadcd96a9e6010d36251651]   <[MessageUid{uid=8}]>
[d578995d125de5e30aadcd96a9e6010d36251651] 
[d578995d125de5e30aadcd96a9e6010d36251651] [ERROR] Errors: 
[d578995d125de5e30aadcd96a9e6010d36251651] [ERROR]   SimpleMessageSearchIndexTest>AbstractMessageSearchIndexTest.searchWithTextShouldReturnMailsWhenBccIsNotAExactMatch:1241 » IllegalArgument
[d578995d125de5e30aadcd96a9e6010d36251651] [ERROR]   SimpleMessageSearchIndexTest>AbstractMessageSearchIndexTest.searchWithTextShouldReturnMailsWhenBccMatches:1232 » IllegalArgument
[d578995d125de5e30aadcd96a9e6010d36251651] [ERROR]   SimpleMessageSearchIndexTest>AbstractMessageSearchIndexTest.searchWithTextShouldReturnMailsWhenCcIsNotAExactMatch:1223 » IllegalArgument
[d578995d125de5e30aadcd96a9e6010d36251651] [ERROR]   SimpleMessageSearchIndexTest>AbstractMessageSearchIndexTest.searchWithTextShouldReturnMailsWhenCcMatches:1214 » IllegalArgument
[d578995d125de5e30aadcd96a9e6010d36251651] [ERROR]   SimpleMessageSearchIndexTest>AbstractMessageSearchIndexTest.searchWithTextShouldReturnMailsWhenFromMatches:1187 » IllegalArgument
[d578995d125de5e30aadcd96a9e6010d36251651] [ERROR]   SimpleMessageSearchIndexTest>AbstractMessageSearchIndexTest.searchWithTextShouldReturnMailsWhenHtmlBodyMatches:1292 » IllegalArgument
[d578995d125de5e30aadcd96a9e6010d36251651] [ERROR]   SimpleMessageSearchIndexTest>AbstractMessageSearchIndexTest.searchWithTextShouldReturnMailsWhenMatches:1196 » IllegalArgument
[d578995d125de5e30aadcd96a9e6010d36251651] [ERROR]   SimpleMessageSearchIndexTest>AbstractMessageSearchIndexTest.searchWithTextShouldReturnMailsWhenTextBodyMatches:1251 » IllegalArgument
[d578995d125de5e30aadcd96a9e6010d36251651] [ERROR]   SimpleMessageSearchIndexTest>AbstractMessageSearchIndexTest.searchWithTextShouldReturnMailsWhenTextBodyMatchesInsensitiveWords:1272 » IllegalArgument
[d578995d125de5e30aadcd96a9e6010d36251651] [ERROR]   SimpleMessageSearchIndexTest>AbstractMessageSearchIndexTest.searchWithTextShouldReturnMailsWhenToIsNotAnExactMatches:1205 » IllegalArgument
[d578995d125de5e30aadcd96a9e6010d36251651] [ERROR]   SimpleMessageSearchIndexTest>AbstractMessageSearchIndexTest.searchWithTextShouldReturnNoMailWhenNotMatching:1178 » IllegalArgument
```

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java [hunk 1]
```diff
@@ -430,15 +430,13 @@ private boolean matchesAddress(SearchQuery.AddressOperator operator, String head
                 AddressList aList = LenientAddressParser.DEFAULT.parseAddressList(value);
                 for (Address address : aList) {
```
Comment (remk) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java]: Maybe  a method should be extracted here ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java [hunk 2]
```diff
@@ -421,26 +421,15 @@ private boolean matches(SearchQuery.HeaderCriterion criterion, MailboxMessage me
      */
     private boolean matchesAddress(SearchQuery.AddressOperator operator, String headerName,
                                    MailboxMessage message) throws MailboxException, IOException {
-        String text = operator.getAddress().toUpperCase(Locale.US);
+        String text = operator.getAddress();
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java]: why?

Comment (trantienduchn) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java]: for example the string `Üsteliğhan` can not be well convert to Upper case with the `Locale.US` but another locale but IDK what is it. I just get the original value, pass it to the String searching method `doesMailboxContains()`. That is the place I convert it to lower case & upper case to increase the chance to match input text.

Comment (mbaechler) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java]: but is it part of this commit?

**Comment by trantienduchn** [comment]
```
[5729103c5b5f3e34eb69d4e769b46f87c3c68ee7] [ERROR] Errors: 
[5729103c5b5f3e34eb69d4e769b46f87c3c68ee7] [ERROR]   Search.testSearchAtomsITALY:63 » InvalidServerResponse 
[5729103c5b5f3e34eb69d4e769b46f87c3c68ee7] Location: /org/apache/...
[5729103c5b5f3e34eb69d4e769b46f87c3c68ee7] [ERROR]   Search.testSearchAtomsKOREA:71 » InvalidServerResponse 
[5729103c5b5f3e34eb69d4e769b46f87c3c68ee7] Location: /org/apache/...
[5729103c5b5f3e34eb69d4e769b46f87c3c68ee7] [ERROR]   Search.testSearchAtomsUS:55 » InvalidServerResponse 
[5729103c5b5f3e34eb69d4e769b46f87c3c68ee7] Location: /org/apache/jam...
[5729103c5b5f3e34eb69d4e769b46f87c3c68ee7] [ERROR]   UidSearch.testSearchAtomsITALY:59 » InvalidServerResponse 
[5729103c5b5f3e34eb69d4e769b46f87c3c68ee7] Location: /org/apac...
[5729103c5b5f3e34eb69d4e769b46f87c3c68ee7] [ERROR]   UidSearch.testSearchAtomsKOREA:66 » InvalidServerResponse 
[5729103c5b5f3e34eb69d4e769b46f87c3c68ee7] Location: /org/apac...
[5729103c5b5f3e34eb69d4e769b46f87c3c68ee7] [ERROR]   UidSearch.testSearchAtomsUS:52 » InvalidServerResponse 
```

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java (hunk 2 above)
Comment (trantienduchn) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java]: okay, do you mean put this change in another commit?

Comment (trantienduchn) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java]: about the `Üsteliğhan` upper case & lower case, I was wrong, just using the uppercase with Locale.US is enough. And just did comparing upper case of two strings like before.

**Comment by trantienduchn** [comment]
```
[3570afb676f0e7caab226114fa05d25cb860d506] [ERROR] Failures: 
[3570afb676f0e7caab226114fa05d25cb860d506] [ERROR]   WebAdminServerIntegrationTest.postShouldDoMigrationAndUpdateToTheLatestVersion:285 1 expectation failed.
[3570afb676f0e7caab226114fa05d25cb860d506] Response body doesn't match expectation.
[3570afb676f0e7caab226114fa05d25cb860d506] Expected: is "{\"version\":7}"
[3570afb676f0e7caab226114fa05d25cb860d506]   Actual: {"version":5}
```
test this please

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java [hunk 3]
```diff
@@ -421,37 +421,54 @@ private boolean matches(SearchQuery.HeaderCriterion criterion, MailboxMessage me
...
-                        }
-                    }
+                AddressList addressList = LenientAddressParser.DEFAULT.parseAddressList(value);
+                if (matchesAddress(addressList, text)) {
+                    return true;
                 }
 
                 // Also try to match against raw header now
-                return value.toUpperCase(Locale.US).contains(text);
+                return value.toUpperCase(Locale.US).contains(text.toUpperCase(Locale.US));
```
Comment (blackheaven) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java]: maybe have a `upCase` method which does the `string.toUpperCase(Local.US)`

Comment (trantienduchn) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java]: not sure about the value it brings.

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**